examples/sugarscape_ig/ss_polars/agents.py

```diff
@@ -25,77 +28,182 @@ def __init__(
+        pl.DataFrame
+            Neighborhood DataFrame
+        """
         neighborhood: pl.DataFrame = self.space.get_neighborhood(
             radius=self["vision"], agents=self, include_center=True
         )
-
         # Join self.space.cells to obtain properties ('sugar') per cell
+
         neighborhood = neighborhood.join(self.space.cells, on=["dim_0", "dim_1"])
 
         # Join self.pos to obtain the agent_id of the center cell
         # TODO: get_neighborhood/get_neighbors should return 'agent_id_center' instead of center position when input is AgentLike
+
         neighborhood = neighborhood.with_columns(
             agent_id_center=neighborhood.join(
                 self.pos,
                 left_on=["dim_0_center", "dim_1_center"],
                 right_on=["dim_0", "dim_1"],
             )["unique_id"]
         )
+        return neighborhood
+
+    def _get_agent_order(self, neighborhood: pl.DataFrame) -> pl.DataFrame:
+        """Get the order of agents based on the original order of agents
+
+        Parameters
+        ----------
+        neighborhood : pl.DataFrame
+            Neighborhood DataFrame
 
+        Returns
+        -------
+        pl.DataFrame
+            DataFrame with 'agent_id_center' and 'agent_order' columns
+        """
         # Order of agents moves based on the original order of agents.
         # The agent in his cell has order 0 (highest)
-        agent_order = neighborhood.unique(
-            subset=["agent_id_center"], keep="first", maintain_order=True
-        ).with_row_count("agent_order")
 
+        return (
+            neighborhood.unique(
+                subset=["agent_id_center"], keep="first", maintain_order=True
+            )
+            .with_row_count("agent_order")
+            .select(["agent_id_center", "agent_order"])
+        )
+
+    def _prepare_neighborhood(
+        self, neighborhood: pl.DataFrame, agent_order: pl.DataFrame
+    ) -> pl.DataFrame:
+        """Prepare the neighborhood DataFrame to find the best moves
+
+        Parameters
+        ----------
+        neighborhood : pl.DataFrame
+            Neighborhood DataFrame
+        agent_order : pl.DataFrame
+            DataFrame with 'agent_id_center' and 'agent_order' columns
+
+        Returns
+        -------
+        pl.DataFrame
+            Prepared neighborhood DataFrame
+        """
         neighborhood = neighborhood.join(agent_order, on="agent_id_center")
 
+        # Add blocking agent order
         neighborhood = neighborhood.join(
             agent_order.select(
                 pl.col("agent_id_center").alias("agent_id"),
                 pl.col("agent_order").alias("blocking_agent_order"),
             ),
             on="agent_id",
-        )
+            how="left",
+        ).rename({"agent_id": "blocking_agent_id"})
 
-        # Filter impossible moves
+        # Filter only possible moves (agent is in his cell, blocking agent has moved before him or there is no blocking agent)
```

Comment:
Why is the previous one, although more concise, not considered favorable?

Reply:
Do you mean the previous implementation?

Reply:
Yes.